File: sdk/python/src/panproto/_migration.py
"""Migration builder and compiled migration wrapper.

Migrations define a mapping between two schemas. Once compiled,
they can efficiently transform records via WASM.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, cast, final

from ._errors import MigrationError, WasmError
from ._msgpack import Packable, pack_migration_mapping, pack_to_wasm, unpack_from_wasm
from ._wasm import WasmHandle, WasmModule, create_handle

if TYPE_CHECKING:
    from collections.abc import Mapping

    from ._schema import BuiltSchema
    from ._types import (
        Edge,
        EdgeWire,
        ExistenceReport,
        GetResult,
        LiftResult,
        MigrationMapping,
        MigrationSpec,
    )


# ---------------------------------------------------------------------------
# MigrationBuilder
# ---------------------------------------------------------------------------


@final
class MigrationBuilder:
    """Immutable fluent builder for constructing panproto migrations.

    Each mutation method returns a *new* :class:`MigrationBuilder` instance.
    Call :meth:`compile` to send the specification to WASM.

    Parameters
    ----------
    src : BuiltSchema
        The source schema for this migration.
    tgt : BuiltSchema
        The target schema for this migration.
    wasm : WasmModule
        The WASM module used for compilation.
    vertex_map : dict[str, str], optional
        Accumulated vertex mappings (default empty).
    edge_map : tuple[tuple[Edge, Edge], ...], optional
        Accumulated edge mappings (default empty).
    resolvers : tuple[tuple[tuple[str, str], Edge], ...], optional
        Accumulated resolver entries (default empty).

    Examples
    --------
    >>> migration = (
    ...     panproto.migration(old_schema, new_schema)
    ...     .map("post", "post")
    ...     .map("post:body", "post:body")
    ...     .compile()
    ... )
    """

    __slots__ = ("_edge_map", "_resolvers", "_src", "_tgt", "_vertex_map", "_wasm")

    def __init__(
        self,
        src: BuiltSchema,
        tgt: BuiltSchema,
        wasm: WasmModule,
        vertex_map: dict[str, str] | None = None,
        edge_map: tuple[tuple[Edge, Edge], ...] | None = None,
        resolvers: tuple[tuple[tuple[str, str], Edge], ...] | None = None,
    ) -> None:
        self._src: BuiltSchema = src
        self._tgt: BuiltSchema = tgt
        self._wasm: WasmModule = wasm
        self._vertex_map: dict[str, str] = dict(vertex_map) if vertex_map else {}
        self._edge_map: tuple[tuple[Edge, Edge], ...] = edge_map if edge_map else ()
        self._resolvers: tuple[tuple[tuple[str, str], Edge], ...] = resolvers if resolvers else ()

    # ------------------------------------------------------------------
    # Mutation helpers
    # ------------------------------------------------------------------

    def map(self, src_vertex: str, tgt_vertex: str) -> MigrationBuilder:
        """Map a source vertex to a target vertex.

        Parameters
        ----------
        src_vertex : str
            Vertex id in the source schema.
        tgt_vertex : str
            Vertex id in the target schema.

        Returns
        -------
        MigrationBuilder
            A new builder with the vertex mapping added.
        """
        return MigrationBuilder(
            self._src,
            self._tgt,
            self._wasm,
            {**self._vertex_map, src_vertex: tgt_vertex},
            self._edge_map,
            self._resolvers,
        )

    def map_edge(self, src_edge: Edge, tgt_edge: Edge) -> MigrationBuilder:
        """Map a source edge to a target edge.

        Parameters
        ----------
        src_edge : Edge
            Edge in the source schema.
        tgt_edge : Edge
            Corresponding edge in the target schema.

        Returns
        -------
        MigrationBuilder
            A new builder with the edge mapping added.
        """
        return MigrationBuilder(
            self._src,
            self._tgt,
            self._wasm,
            self._vertex_map,
            (*self._edge_map, (src_edge, tgt_edge)),
            self._resolvers,
        )

    def resolve(
        self,
        src_kind: str,
        tgt_kind: str,
        resolved_edge: Edge,
    ) -> MigrationBuilder:
        """Add a resolver for ancestor-contraction ambiguity.

        When a migration contracts nodes and the resulting edge between
        two vertex kinds is ambiguous, a resolver specifies which edge
        to use.

        Parameters
        ----------
        src_kind : str
            Source vertex kind in the contracted pair.
        tgt_kind : str
            Target vertex kind in the contracted pair.
        resolved_edge : Edge
            The edge to use when resolving this (src_kind, tgt_kind) pair.

        Returns
        -------
        MigrationBuilder
            A new builder with the resolver added.
        """
        return MigrationBuilder(
            self._src,
            self._tgt,
            self._wasm,
            self._vertex_map,
            self._edge_map,
            (*self._resolvers, ((src_kind, tgt_kind), resolved_edge)),
        )

    def to_spec(self) -> MigrationSpec:
        """Return the current migration specification.

        Returns
        -------
        MigrationSpec
            All accumulated vertex mappings, edge mappings, and resolvers.
        """
        return {
            "vertex_map": dict(self._vertex_map),
            "edge_map": list(self._edge_map),
            "resolvers": list(self._resolvers),
        }

    def compile(self) -> CompiledMigration:
        """Compile the migration for fast per-record application.

        Sends the migration specification to WASM. The resulting
        :class:`CompiledMigration` can transform records via
        :meth:`~CompiledMigration.lift`, :meth:`~CompiledMigration.get`,
        and :meth:`~CompiledMigration.put`.

        Returns
        -------
        CompiledMigration
            A compiled migration ready for record transformation.

        Raises
        ------
        MigrationError
            If WASM rejects the migration specification.
        """
        mapping = _build_wasm_mapping(self._vertex_map, self._edge_map, self._resolvers)
        mapping_bytes = pack_migration_mapping(mapping)

        try:
            raw_handle = self._wasm.compile_migration(
                self._src.wasm_handle.id,
                self._tgt.wasm_handle.id,
                mapping_bytes,
            )
        except Exception as exc:
            raise MigrationError(f"Failed to compile migration: {exc}") from exc

        handle = create_handle(raw_handle, self._wasm)
        return CompiledMigration(handle, self._wasm, self.to_spec())


# ---------------------------------------------------------------------------
# CompiledMigration
# ---------------------------------------------------------------------------


@final
class CompiledMigration:
    """A compiled migration that can efficiently transform records via WASM.

    Implements the context-manager protocol for automatic cleanup.

    Parameters
    ----------
    handle : WasmHandle
        WASM handle returned by ``compile_migration``.
    wasm : WasmModule
        The owning WASM module.
    spec : MigrationSpec
        The specification used to produce this migration (for introspection).

    Examples
    --------
    >>> with compiled_migration as m:
    ...     result = m.lift({"text": "hello"})
    """

    __slots__ = ("_handle", "_spec", "_wasm")

    def __init__(
        self,
        handle: WasmHandle,
        wasm: WasmModule,
        spec: MigrationSpec,
    ) -> None:
        self._handle: WasmHandle = handle
        self._wasm: WasmModule = wasm
        self._spec: MigrationSpec = spec

    # ------------------------------------------------------------------
    # Internal accessors
    # ------------------------------------------------------------------

    @property
    def wasm_handle(self) -> WasmHandle:
        """The underlying WASM handle (internal use only)."""
        return self._handle

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    @property
    def spec(self) -> MigrationSpec:
        """The migration specification used to build this migration.

        Returns
        -------
        MigrationSpec
        """
        return self._spec

    def lift(self, record: Packable) -> LiftResult:
        """Transform a record using this migration (forward direction).

        Parameters
        ----------
        record : Packable
            The input record to transform. Must be MessagePack-serialisable.

        Returns
        -------
        LiftResult
            A dict with a single ``"data"`` key containing the
            transformed record.

        Raises
        ------
        WasmError
            If the WASM call fails.
        """
        input_bytes = pack_to_wasm(record)
        try:
            output_bytes = self._wasm.lift_record(self._handle.id, input_bytes)
        except Exception as exc:
            raise WasmError(f"lift_record failed: {exc}") from exc
        data = unpack_from_wasm(output_bytes)
        return {"data": data}

    def get(self, record: Packable) -> GetResult:
        """Bidirectional get: extract view and complement from a record.

        The complement captures data discarded by the forward projection,
        enabling lossless round-tripping via :meth:`put`.

        Parameters
        ----------
        record : Packable
            The input record. Must be MessagePack-serialisable.

        Returns
        -------
        GetResult
            A dict with ``"view"`` (the projected record) and
            ``"complement"`` (opaque ``bytes`` for round-tripping).

        Raises
        ------
        WasmError
            If the WASM call fails.
        """
        input_bytes = pack_to_wasm(record)
        try:
            output_bytes = self._wasm.get_record(self._handle.id, input_bytes)
        except Exception as exc:
            raise WasmError(f"get_record failed: {exc}") from exc

        raw = unpack_from_wasm(output_bytes)
        result = cast("Mapping[str, Packable]", raw)
        complement_raw = result.get("complement", b"")
        complement = (
            bytes(complement_raw)  # type: ignore[arg-type]
            if not isinstance(complement_raw, bytes)
            else complement_raw
        )
        return {"view": result.get("view"), "complement": complement}

    def put(self, view: Packable, complement: bytes) -> LiftResult:
        """Bidirectional put: restore a full record from view and complement.

        Parameters
        ----------
        view : Packable
            The (possibly modified) projected view.
        complement : bytes
            The opaque complement bytes from a prior :meth:`get` call.

        Returns
        -------
        LiftResult
            A dict with a single ``"data"`` key containing the
            restored full record.

        Raises
        ------
        WasmError
            If the WASM call fails.
        """
        view_bytes = pack_to_wasm(view)
        try:
            output_bytes = self._wasm.put_record(self._handle.id, view_bytes, complement)
        except Exception as exc:
            raise WasmError(f"put_record failed: {exc}") from exc
        data = unpack_from_wasm(output_bytes)
        return {"data": data}

    # ------------------------------------------------------------------
    # Context manager / cleanup
    # ------------------------------------------------------------------

    def close(self) -> None:
        """Release the WASM-side compiled migration resource.

        Safe to call multiple times; subsequent calls are no-ops.
        """
        self._handle.close()

    def __enter__(self) -> CompiledMigration:
        return self

    def __exit__(self, *_: object) -> None:
        self.close()


# ---------------------------------------------------------------------------
# Module-level functions
# ---------------------------------------------------------------------------


def check_existence(
    src: BuiltSchema,
    tgt: BuiltSchema,
    spec: MigrationSpec,
    wasm: WasmModule,
) -> ExistenceReport:
    """Check existence conditions for a migration between two schemas.

    Parameters
    ----------
    src : BuiltSchema
        The source schema.
    tgt : BuiltSchema
        The target schema.
    spec : MigrationSpec
        The migration specification to validate.
    wasm : WasmModule
        The WASM module.

    Returns
    -------
    ExistenceReport
        Validation result with ``valid`` flag and any ``errors``.
    """
    mapping = _build_wasm_mapping(
        spec["vertex_map"],
        spec["edge_map"],
        spec["resolvers"],
    )
    mapping_bytes = pack_migration_mapping(mapping)

    result_bytes = wasm.check_existence(
        src.wasm_handle.id,
        tgt.wasm_handle.id,
        mapping_bytes,
    )
    return cast("ExistenceReport", unpack_from_wasm(result_bytes))


def compose_migrations(
    m1: CompiledMigration,
    m2: CompiledMigration,
    wasm: WasmModule,
) -> CompiledMigration:
    """Compose two compiled migrations into a single migration.

    The resulting migration is equivalent to applying *m1* first, then *m2*.

    Parameters
    ----------
    m1 : CompiledMigration
        First migration (applied first).
    m2 : CompiledMigration
        Second migration (applied second).
    wasm : WasmModule
        The WASM module.

    Returns
    -------
    CompiledMigration
        A new compiled migration representing ``m2 . m1``.

    Raises
    ------
    MigrationError
        If WASM composition fails.
    """
    try:
        raw_handle = wasm.compose_migrations(
            m1.wasm_handle.id,
            m2.wasm_handle.id,
        )
    except Exception as exc:
        raise MigrationError(f"Failed to compose migrations: {exc}") from exc

    handle = create_handle(raw_handle, wasm)

    # Compose vertex maps: A->B in m1 + B->C in m2 = A->C.
    composed_vertex_map: dict[str, str] = {}
    for src_v, intermediate in m1.spec["vertex_map"].items():
        final_v = m2.spec["vertex_map"].get(intermediate, intermediate)
        composed_vertex_map[src_v] = final_v

    composed_spec: MigrationSpec = {
        "vertex_map": composed_vertex_map,
        "edge_map": [*m1.spec["edge_map"], *m2.spec["edge_map"]],
        "resolvers": [*m1.spec["resolvers"], *m2.spec["resolvers"]],
    }

    return CompiledMigration(handle, wasm, composed_spec)


# ---------------------------------------------------------------------------
# Private helpers
# ---------------------------------------------------------------------------


def _edge_to_wire(e: Edge) -> EdgeWire:
    """Convert an Edge dataclass to wire format.

    Parameters
    ----------
    e : Edge
        The edge to convert.

    Returns
    -------
    EdgeWire
        Wire-format representation of the edge.
    """
    return {
        "src": e["src"],
        "tgt": e["tgt"],
        "kind": e["kind"],
        "name": e.get("name"),
    }


def _build_wasm_mapping(
    vertex_map: dict[str, str] | Mapping[str, str],
    edge_map: list[tuple[Edge, Edge]] | tuple[tuple[Edge, Edge], ...],
    resolvers: (list[tuple[tuple[str, str], Edge]] | tuple[tuple[tuple[str, str], Edge], ...]),
) -> MigrationMapping:
    """Build a WASM-compatible migration mapping from builder state.

    Parameters
    ----------
    vertex_map : dict[str, str] | Mapping[str, str]
        Source vertex id to target vertex id mappings.
    edge_map : list[tuple[Edge, Edge]] | tuple[tuple[Edge, Edge], ...]
        Pairs of (source edge, target edge).
    resolvers : list[tuple[tuple[str, str], Edge]] | tuple[tuple[tuple[str, str], Edge], ...]
        Resolver entries: ``((src_kind, tgt_kind), resolved_edge)``.

    Returns
    -------
    MigrationMapping
        Wire-format mapping ready for MessagePack serialization.
    """
    return {
        "vertex_map": dict(vertex_map),
        "edge_map": [(_edge_to_wire(s), _edge_to_wire(t)) for s, t in edge_map],
        "hyper_edge_map": {},
        "label_map": [],
        "resolver": [(list(key), _edge_to_wire(edge)) for key, edge in resolvers],
    }

File: sdk/typescript/src/migration.ts
/**
 * Migration builder and compiled migration wrapper.
 *
 * Migrations define a mapping between two schemas. Once compiled,
 * they can efficiently transform records via WASM.
 *
 * @module
 */

import type {
  WasmModule,
  Edge,
  LiftResult,
  GetResult,
  ExistenceReport,
  MigrationSpec,
} from './types.js';
import { MigrationError, WasmError } from './types.js';
import { WasmHandle, createHandle } from './wasm.js';
import { packToWasm, unpackFromWasm, packMigrationMapping } from './msgpack.js';
import type { BuiltSchema } from './schema.js';

/**
 * Fluent builder for constructing migrations between two schemas.
 *
 * Each mutation method returns a new `MigrationBuilder` instance,
 * leaving the original unchanged.
 *
 * @example
 * ```typescript
 * const migration = panproto.migration(oldSchema, newSchema)
 *   .map('post', 'post')
 *   .map('post:body', 'post:body')
 *   .mapEdge(oldEdge, newEdge)
 *   .compile();
 * ```
 */
export class MigrationBuilder {
  readonly #src: BuiltSchema;
  readonly #tgt: BuiltSchema;
  readonly #wasm: WasmModule;
  readonly #vertexMap: ReadonlyMap<string, string>;
  readonly #edgeMap: readonly [Edge, Edge][];
  readonly #resolvers: readonly [[string, string], Edge][];

  constructor(
    src: BuiltSchema,
    tgt: BuiltSchema,
    wasm: WasmModule,
    vertexMap: ReadonlyMap<string, string> = new Map(),
    edgeMap: readonly [Edge, Edge][] = [],
    resolvers: readonly [[string, string], Edge][] = [],
  ) {
    this.#src = src;
    this.#tgt = tgt;
    this.#wasm = wasm;
    this.#vertexMap = vertexMap;
    this.#edgeMap = edgeMap;
    this.#resolvers = resolvers;
  }

  /**
   * Map a source vertex to a target vertex.
   *
   * @param srcVertex - Vertex id in the source schema
   * @param tgtVertex - Vertex id in the target schema
   * @returns A new builder with the mapping added
   */
  map(srcVertex: string, tgtVertex: string): MigrationBuilder {
    const newMap = new Map(this.#vertexMap);
    newMap.set(srcVertex, tgtVertex);

    return new MigrationBuilder(
      this.#src,
      this.#tgt,
      this.#wasm,
      newMap,
      this.#edgeMap,
      this.#resolvers,
    );
  }

  /**
   * Map a source edge to a target edge.
   *
   * @param srcEdge - Edge in the source schema
   * @param tgtEdge - Edge in the target schema
   * @returns A new builder with the edge mapping added
   */
  mapEdge(srcEdge: Edge, tgtEdge: Edge): MigrationBuilder {
    return new MigrationBuilder(
      this.#src,
      this.#tgt,
      this.#wasm,
      this.#vertexMap,
      [...this.#edgeMap, [srcEdge, tgtEdge]],
      this.#resolvers,
    );
  }

  /**
   * Add a resolver for ancestor contraction ambiguity.
   *
   * When a migration contracts nodes and the resulting edge between
   * two vertex kinds is ambiguous, a resolver specifies which edge to use.
   *
   * @param srcKind - Source vertex kind in the contracted pair
   * @param tgtKind - Target vertex kind in the contracted pair
   * @param resolvedEdge - The edge to use for this pair
   * @returns A new builder with the resolver added
   */
  resolve(srcKind: string, tgtKind: string, resolvedEdge: Edge): MigrationBuilder {
    return new MigrationBuilder(
      this.#src,
      this.#tgt,
      this.#wasm,
      this.#vertexMap,
      this.#edgeMap,
      [...this.#resolvers, [[srcKind, tgtKind], resolvedEdge]],
    );
  }

  /**
   * Get the current migration specification.
   *
   * @returns The migration spec with all accumulated mappings
   */
  toSpec(): MigrationSpec {
    return {
      vertexMap: Object.fromEntries(this.#vertexMap),
      edgeMap: [...this.#edgeMap],
      resolvers: [...this.#resolvers],
    };
  }

  /**
   * Invert a bijective migration.
   *
   * Serializes the current mapping to MessagePack and calls the
   * `invert_migration` WASM entry point. Returns a new MigrationSpec
   * representing the inverted migration.
   *
   * @returns The inverted migration specification
   * @throws {@link MigrationError} if the migration is not bijective or inversion fails
   */
  invert(): MigrationSpec {
    const edgeMap = new Map(
      this.#edgeMap.map(([src, tgt]) => [
        { src: src.src, tgt: src.tgt, kind: src.kind, name: src.name ?? null },
        { src: tgt.src, tgt: tgt.tgt, kind: tgt.kind, name: tgt.name ?? null },
      ] as const),
    );
    const resolver = new Map(
      this.#resolvers.map(([[s, t], e]) => [
        [s, t] as const,
        { src: e.src, tgt: e.tgt, kind: e.kind, name: e.name ?? null },
      ] as const),
    );
    const mapping = packMigrationMapping({
      vertex_map: Object.fromEntries(this.#vertexMap),
      edge_map: edgeMap,
      hyper_edge_map: {},
      label_map: new Map(),
      resolver,
    });

    try {
      const resultBytes = this.#wasm.exports.invert_migration(
        mapping,
        this.#src._handle.id,
        this.#tgt._handle.id,
      );
      return unpackFromWasm<MigrationSpec>(resultBytes);
    } catch (error) {
      throw new MigrationError(
        `Failed to invert migration: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Compile the migration for fast per-record application.
   *
   * Sends the migration specification to WASM for compilation.
   * The resulting `CompiledMigration` can be used to transform records.
   *
   * @returns A compiled migration ready for record transformation
   * @throws {@link MigrationError} if compilation fails
   */
  compile(): CompiledMigration {
    const edgeMap = new Map(
      this.#edgeMap.map(([src, tgt]) => [
        { src: src.src, tgt: src.tgt, kind: src.kind, name: src.name ?? null },
        { src: tgt.src, tgt: tgt.tgt, kind: tgt.kind, name: tgt.name ?? null },
      ] as const),
    );
    const resolver = new Map(
      this.#resolvers.map(([[s, t], e]) => [
        [s, t] as const,
        { src: e.src, tgt: e.tgt, kind: e.kind, name: e.name ?? null },
      ] as const),
    );
    const mapping = packMigrationMapping({
      vertex_map: Object.fromEntries(this.#vertexMap),
      edge_map: edgeMap,
      hyper_edge_map: {},
      label_map: new Map(),
      resolver,
    });

    try {
      const rawHandle = this.#wasm.exports.compile_migration(
        this.#src._handle.id,
        this.#tgt._handle.id,
        mapping,
      );

      const handle = createHandle(rawHandle, this.#wasm);
      return new CompiledMigration(handle, this.#wasm, this.toSpec());
    } catch (error) {
      throw new MigrationError(
        `Failed to compile migration: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }
}

/**
 * A compiled migration that can efficiently transform records via WASM.
 *
 * Implements `Disposable` for automatic cleanup of the WASM resource.
 */
export class CompiledMigration implements Disposable {
  readonly #handle: WasmHandle;
  readonly #wasm: WasmModule;
  readonly #spec: MigrationSpec;

  constructor(handle: WasmHandle, wasm: WasmModule, spec: MigrationSpec) {
    this.#handle = handle;
    this.#wasm = wasm;
    this.#spec = spec;
  }

  /** The WASM handle for this migration. Internal use only. */
  get _handle(): WasmHandle {
    return this.#handle;
  }

  /** The WASM module. Internal use only. */
  get _wasm(): WasmModule {
    return this.#wasm;
  }

  /** The migration specification used to build this migration. */
  get spec(): MigrationSpec {
    return this.#spec;
  }

  /**
   * Transform a record using this migration (forward direction).
   *
   * Accepts either a raw JS object (which will be msgpack-encoded) or
   * an `Instance` (whose pre-encoded bytes are passed directly to WASM).
   *
   * @param record - The input record or Instance to transform
   * @returns The transformed record
   * @throws {@link WasmError} if the WASM call fails
   */
  lift(record: unknown): LiftResult {
    // If record has _bytes, treat it as an Instance and pass bytes directly
    const inputBytes = (record && typeof record === 'object' && '_bytes' in record)
      ? (record as { _bytes: Uint8Array })._bytes
      : packToWasm(record);

    try {
      const outputBytes = this.#wasm.exports.lift_record(
        this.#handle.id,
        inputBytes,
      );
      const data = unpackFromWasm(outputBytes);
      return { data, _rawBytes: outputBytes };
    } catch (error) {
      throw new WasmError(
        `lift_record failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Bidirectional get: extract view and complement from a record.
   *
   * The complement captures data discarded by the forward projection,
   * enabling lossless round-tripping via `put()`.
   *
   * @param record - The input record
   * @returns The projected view and opaque complement bytes
   * @throws {@link WasmError} if the WASM call fails
   */
  get(record: unknown): GetResult {
    const inputBytes = (record && typeof record === 'object' && '_bytes' in record)
      ? (record as { _bytes: Uint8Array })._bytes
      : packToWasm(record);

    try {
      const outputBytes = this.#wasm.exports.get_record(
        this.#handle.id,
        inputBytes,
      );
      const result = unpackFromWasm<{ view: unknown; complement: Uint8Array }>(outputBytes);
      return {
        view: result.view,
        complement: result.complement instanceof Uint8Array
          ? result.complement
          : new Uint8Array(result.complement as ArrayBuffer),
      };
    } catch (error) {
      throw new WasmError(
        `get_record failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Bidirectional put: restore a full record from a modified view and complement.
   *
   * @param view - The (possibly modified) projected view
   * @param complement - The complement from a prior `get()` call
   * @returns The restored full record
   * @throws {@link WasmError} if the WASM call fails
   */
  put(view: unknown, complement: Uint8Array): LiftResult {
    const viewBytes = (view && typeof view === 'object' && '_bytes' in view)
      ? (view as { _bytes: Uint8Array })._bytes
      : packToWasm(view);

    try {
      const outputBytes = this.#wasm.exports.put_record(
        this.#handle.id,
        viewBytes,
        complement,
      );
      const data = unpackFromWasm(outputBytes);
      return { data };
    } catch (error) {
      throw new WasmError(
        `put_record failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /** Release the WASM-side compiled migration resource. */
  [Symbol.dispose](): void {
    this.#handle[Symbol.dispose]();
  }
}

/**
 * Check existence conditions for a migration between two schemas.
 *
 * @param src - Source schema handle
 * @param tgt - Target schema handle
 * @param spec - The migration specification
 * @param wasm - The WASM module
 * @returns The existence report
 */
export function checkExistence(
  protoHandle: number,
  src: BuiltSchema,
  tgt: BuiltSchema,
  spec: MigrationSpec,
  wasm: WasmModule,
): ExistenceReport {
  const edgeMap = new Map(
    spec.edgeMap.map(([s, t]) => [
      { src: s.src, tgt: s.tgt, kind: s.kind, name: s.name ?? null },
      { src: t.src, tgt: t.tgt, kind: t.kind, name: t.name ?? null },
    ] as const),
  );
  const resolver = new Map(
    spec.resolvers.map(([[s, t], e]) => [
      [s, t] as const,
      { src: e.src, tgt: e.tgt, kind: e.kind, name: e.name ?? null },
    ] as const),
  );
  const mapping = packMigrationMapping({
    vertex_map: spec.vertexMap,
    edge_map: edgeMap,
    hyper_edge_map: {},
    label_map: new Map(),
    resolver,
  });

  const resultBytes = wasm.exports.check_existence(
    protoHandle,
    src._handle.id,
    tgt._handle.id,
    mapping,
  );

  return unpackFromWasm<ExistenceReport>(resultBytes);
}

/**
 * Compose two compiled migrations into a single migration.
 *
 * @param m1 - First migration (applied first)
 * @param m2 - Second migration (applied second)
 * @param wasm - The WASM module
 * @returns A new compiled migration representing m2 . m1
 * @throws {@link MigrationError} if composition fails
 */
export function composeMigrations(
  m1: CompiledMigration,
  m2: CompiledMigration,
  wasm: WasmModule,
): CompiledMigration {
  try {
    const rawHandle = wasm.exports.compose_migrations(
      m1._handle.id,
      m2._handle.id,
    );
    const handle = createHandle(rawHandle, wasm);

    // Compose vertex maps: if m1 maps A->B and m2 maps B->C, composed maps A->C.
    // Vertices in m1 whose target is not remapped by m2 pass through unchanged.
    const composedVertexMap: Record<string, string> = {};
    for (const [src, intermediate] of Object.entries(m1.spec.vertexMap)) {
      const final_ = m2.spec.vertexMap[intermediate];
      composedVertexMap[src] = final_ ?? intermediate;
    }

    // Concatenate edge maps and resolvers from both migrations. The WASM side
    // performs the actual composition; this spec is a best-effort reconstruction
    // for introspection purposes.
    const composedSpec: MigrationSpec = {
      vertexMap: composedVertexMap,
      edgeMap: [...m1.spec.edgeMap, ...m2.spec.edgeMap],
      resolvers: [...m1.spec.resolvers, ...m2.spec.resolvers],
    };

    return new CompiledMigration(handle, wasm, composedSpec);
  } catch (error) {
    throw new MigrationError(
      `Failed to compose migrations: ${error instanceof Error ? error.message : String(error)}`,
      { cause: error },
    );
  }
}
